### DISCUSSION TITLE
Incorrect answer

### DISCUSSION THREAD

**Discussion by chenYanbin0218** [ERGO-Code/HiGHS#1878] [Q&A]
When I solve the following simple questions using HiGHS through pyoptinterface, the output is different from the answer using Gurobi
![微信截图_20240816210827](https://github.com/user-attachments/assets/14a86a6d-21bc-4e37-a487-a6fdff66b9c2)

 ### Using  HiGHS
```
import pyoptinterface as poi
from pyoptinterface import highs
model = highs.Model()
x = model.add_variable(domain=poi.VariableDomain.Binary, name="x")
y = model.add_variable(domain=poi.VariableDomain.Binary, name="y")
z = model.add_variable(lb=0,ub=100, name="z")
con1=model.add_linear_constraint(2*y+3*z, poi.ConstraintSense.LessEqual, 12, name="con1")
con2=model.add_linear_constraint(x+y, poi.ConstraintSense.LessEqual, 1, name="con2")
obj =3*y+x*z
model.set_objective(obj, poi.ObjectiveSense.Maximize)
model.optimize()

print("x = ", model.get_value(x))
print("y = ", model.get_value(y))
print("z = ", model.get_value(z))
print("obj = ", model.get_value(obj))
```

### Output of   HiGHS
```
Presolving model
2 rows, 1 cols, 2 nonzeros  0s
0 rows, 0 cols, 0 nonzeros  0s
Presolve: Optimal

Solving report
  Status            Optimal
  Primal bound      3
  Dual bound        3
  Gap               0% (tolerance: 0.01%)
  Solution status   feasible
                    3 (objective)
                    0 (bound viol.)
                    0 (int. viol.)
                    0 (row viol.)
  Timing            0.00 (total)
                    0.00 (presolve)
                    0.00 (postsolve)
  Nodes             0
  LP iterations     0 (total)
                    0 (strong br.)
                    0 (separation)
                    0 (heuristics)
x =  0.0
y =  1.0
z =  0.0
obj =  3.0
```
 
### Using  Gurobi
```
import pyoptinterface as poi
from pyoptinterface import gurobi
 model = gurobi.Model()
x = model.add_variable(domain=poi.VariableDomain.Binary, name="x")
y = model.add_variable(domain=poi.VariableDomain.Binary, name="y")
z = model.add_variable(lb=0,ub=100, name="z")
con1=model.add_linear_constraint(2*y+3*z, poi.ConstraintSense.LessEqual, 12, name="con1")
con2=model.add_linear_constraint(x+y, poi.ConstraintSense.LessEqual, 1, name="con2")
obj =3*y+x*z
model.set_objective(obj, poi.ObjectiveSense.Maximize)
model.optimize()

print("x = ", model.get_value(x))
print("y = ", model.get_value(y))
print("z = ", model.get_value(z))
print("obj = ", model.get_value(obj))
```

### Output of   Gurobi
```
Optimal solution found (tolerance 1.00e-04)
Best objective 4.000000000000e+00, best bound 4.000000000000e+00, gap 0.0000%
x =  1.0
y =  0.0
z =  4.0
obj =  4.0
```

**Comment by jajhall** [comment]
This is because your problem is mixed-integer quadratic programming, and HiGHS can't solve such problems -
[README.md](https://github.com/ERGO-Code/HiGHS/blob/master/README.md)

At one point, HiGHS used to ignore the Hessian matrix and just solve the mixed-integer linear programming (MIP) problem, but I'm struggling to find a version in which that happens! Now, all that happens is that HiGHS gives the message

> ERROR:   Cannot solve MIQP problems with HiGHS

What version of HiGHS is used by `pyoptinterface`?

Since the HiGHS solution that you quote corresponds to ignoring the Hessian, could it be that `pyoptinterface` doesn't pass the Hessian to HiGHS once it knows that there are integer variables?

**Reply by chenYanbin0218** [reply]
Thank you very much for your reply! The version is 1.7.1 and I think it's pyoptinerface that results such situation, missing the error.

**Reply by metab0t** [reply]
@jajhall I want to ensure that HiGHS accepts Hessian matrix in the lower triangular form or upper triangular form?

The docstring in C API says HiGHS accepts upper triangular form but the test codes in repo show that we should use lower triangle.

https://github.com/ERGO-Code/HiGHS/blob/5ce7a27531a7f4166ee5a8343169a1014febb41a/src/interfaces/highs_c_api.h#L532-L533

**Reply by jajhall** [reply]
> @jajhall I want to ensure that HiGHS accepts Hessian matrix in the lower triangular form or upper triangular form?
> 
> The docstring in C API says HiGHS accepts upper triangular form but the test codes in repo show that we should use lower triangle.
> 
> https://github.com/ERGO-Code/HiGHS/blob/5ce7a27531a7f4166ee5a8343169a1014febb41a/src/interfaces/highs_c_api.h#L532-L533

Indeed, this should say "lower triangular component". Fixed by #1879

**Reply by metab0t** [reply]
Thanks! I have pushed a fix of PyOptInterface to use the lower triangular of Hessian.
